# <center> Homework 1 (Topics 1-4: Data Structures, Indexing, Cleaning Data, Math)

In [1]:
import pandas as pd
import numpy as np

#### Task 1. 
Задание:
1. Создайте Series из последовательности 15 значений, равномерно разбивающих отрезок [0, 20]. 
2. Определите отношение элементов полученной серии к их предыдущим элементам (\*). 
3. В результате необходимо получить среднее полученного вектора, оставив в нём только те значения, которые не более чем 1.5 (\**). 

Выберите из ответов тот, который максимально близок к полученному (c точки зрения абсолютной разницы).

#### Варианты ответов:
    - 1) 1.24
    - 2) 1.18
    - 3) 0.71
    - 4) 1.13


##### Пояснения:
(\*) Если было бы необходимо найти последовательность из 3-х значений,  
равномерно разбивающих отрезок [0,1], то это были бы значения [0, 0.5, 1]. 

(\**) Если был бы дан список элементов a = [1,2,3,12], отношения элементов к предыдущим будут равны [NaN, 2, 1.5, 4].  
А на последнем этапе в таком примере останется только [1.5] и среднее значение будет также 1.5.

In [15]:
s = pd.Series(np.linspace(0, 20, 15))
s

0      0.000000
1      1.428571
2      2.857143
3      4.285714
4      5.714286
5      7.142857
6      8.571429
7     10.000000
8     11.428571
9     12.857143
10    14.285714
11    15.714286
12    17.142857
13    18.571429
14    20.000000
dtype: float64

In [25]:
division = pd.Series(np.divide(s.values[1:],s.values[:-1]))
division

C:\Users\ogloc\AppData\Local\Temp\ipykernel_3184\2846618547.py:1: RuntimeWarning: divide by zero encountered in divide
  division = pd.Series(np.divide(s.values[1:],s.values[:-1]))


0          inf
1     2.000000
2     1.500000
3     1.333333
4     1.250000
5     1.200000
6     1.166667
7     1.142857
8     1.125000
9     1.111111
10    1.100000
11    1.090909
12    1.083333
13    1.076923
dtype: float64

In [26]:
vec_mean = division[division <= 1.5].mean()
vec_mean

1.181677812927813

2) 1.18

#### Task 2. 
Выберите все верные ответы касательно следующих 3-х Series:
    - pd.Series('abcde'); (1) 
    - pd.Series(['abcde']); (2) 
    - pd.Series(list('abcde')); (3)
    - pd.Series("abcde"); (4)

#### Вопросы:
    - 1) Серия (1) совпадает с серией (2), так как в каждом из случаев серия создаётся из списка строк 
    - 2) Серия (2) совпадает с серией (3), так как в каждом из случаев серия создаётся из списка символов 
    - 3) Серия (1) не совпадает с серией (4), так как в (4) используются двойные кавычки "" вместо одинарных '' 

In [31]:
task2_1 = pd.Series('abcde')
task2_2 = pd.Series(['abcde'])
task2_3 = pd.Series(list('abcde'))
task2_4 = pd.Series("abcde")
print(task2_1)
print(task2_2)
print(task2_3)
print(task2_4)


0    abcde
dtype: object
0    abcde
dtype: object
0    a
1    b
2    c
3    d
4    e
dtype: object
0    abcde
dtype: object


  - 1) Серия (1) совпадает с серией (2), так как в каждом из случаев серия создаётся из списка строк 
  

#### Task 3. 
Для заданного ниже объекта s типа Series проделайте следующее:
1. Создайте новый Series, значения которого совпадают со значениями s, а индексы - целочисленные значения от 2 до 12. 
2. Выберите из s элементы с индексами 3 и 5, после чего просуммируйте их, сохранив результат (1). 
3. Выберите из s только целочисленные элементы и вычислите их дисперсию (2). (\*)

Все полученные результаты округлите до 2-х знаков после запятой.

#### Выберите все верные пункты:
    - 1) Ответ (1) - 642.52
    - 2) Ответ (1) - 91.78
    - 3) Ответ (1) - не может быть определён
    - 4) Ответ (2) - 57591.19
    - 5) Ответ (2) - 210.12
    - 6) Ответ (2) - не может быть определён

##### Пояснения:
(\*) Целочисленные значения - значения, имеющие тип int.

In [32]:
s = pd.Series(data=['1', 2, 3.1, 'hi!', 5, -512, 12.42, 'sber', 10.10, 98], 
              index=range(6, 26, 2))
s

6         1
8         2
10      3.1
12      hi!
14        5
16     -512
18    12.42
20     sber
22     10.1
24       98
dtype: object

In [117]:
s2 = s.copy()
s2 = s2.set_axis(np.arange(2,12))
s2

2         1
3         2
4       3.1
5       hi!
6         5
7      -512
8     12.42
9      sber
10     10.1
11       98
dtype: object

In [65]:
sum_3_and_5 = s2.loc[3] + s2.loc[5]
sum_3_and_5


TypeError: unsupported operand type(s) for +: 'int' and 'str'

    - 3) Ответ (1) - не может быть определён
    

In [129]:
s2 = s2[s2.apply(lambda x: type(x) == int)]


In [132]:
np.var(s2)

57591.1875

    - 4) Ответ (2) - 57591.19

#### Task 4. 
1. Сгенерируйте Series из 100 значений нормально распределённой СВ (np.random.normal с дефолтными параметрами - нулевым средним и единичной дисперсией).
2. Возведите каждое значение серии в 3 степень, а значения индекса увеличьте в 3 раза. 
3. Ответьте на следующие вопросы через запятую (без пробелов) (\*)
    - 1) Выведите сумму элементов, строго меньших 2.6, имеющих нечётные значения индекса.
    - 2) Выведите количество значений серии меньше нуля. 

#### Пояснения:
(\*) Если получились ответы 3, 4.32, то в форму необходимо внести их в виде "3,4.32". То есть вещественные числа необходимо разделять точками. Не забудьте про фиксированный seed (его менять не нужно)!

In [142]:
np.random.seed(242)
s4 = pd.Series(np.random.normal(0, 1, 100))
s4

0    -0.357519
1     0.148448
2     0.993531
3     1.838968
4    -0.744026
        ...   
95    0.950296
96    0.737315
97   -0.739985
98   -1.376260
99    0.052260
Length: 100, dtype: float64

In [148]:
s4_2 = s4**3
s4_2 = s4_2.set_axis(s4_2.index*3)
s4_2

0     -0.045698
3      0.003271
6      0.980718
9      6.219024
12    -0.411874
         ...   
285    0.858176
288    0.400829
291   -0.405199
294   -2.606764
297    0.000143
Length: 100, dtype: float64

In [160]:
s4_2[(s4_2.index % 2 == 1) & (s4_2 < 2.6)].sum()

-15.150748101821666

In [158]:
(s4_2 < 0).sum()

51

-15.15,51

### Информация для последующих заданий
- Для всех последующих заданий будем использовать обезличенные транзакционные банковские данные. Для этого считайте в переменные tr_mcc_codes, tr_types, transactions и gender_train из одноимённых таблиц из папки data. Для таблицы transactions используйте только первые n=1000000 строк. Обратите внимание на разделители внутри каждого из файлов - они могут различаться!

In [237]:
tr_mcc_codes = pd.read_csv(filepath_or_buffer = "../data/tr_mcc_codes.csv",
                    sep = ';')
tr_types = pd.read_csv(filepath_or_buffer = "../data/tr_types.csv",
                    sep = ';',
                    index_col = 'tr_type')
transactions = pd.read_csv(filepath_or_buffer = "../data/transactions.csv",
                    sep = ',',
                    nrows = 1000000)
gender_train = pd.read_csv(filepath_or_buffer = "../data/gender_train.csv",
                    sep = ',')


In [171]:
tr_mcc_codes

mcc_code                                    mcc_description
0         742                                Ветеринарные услуги
1        1711  Генеральные подрядчики по вентиляции, теплосна...
2        1731                        Подрядчики по электричеству
3        1799  Подрядчики, специализированная торговля — нигд...
4        2741           Разнообразные издательства/печатное дело
..        ...                                                ...
179      9211  Судовые выплаты, включая алименты и детскую по...
180      9222                                             Штрафы
181      9311                                  Налоговые платежи
182      9399  Правительственные услуги, нигде ранее не класс...
183      9402             Почтовые услуги — только правительство

[184 rows x 2 columns]

In [172]:
tr_types

tr_type                                     tr_description
0       3200  Плата за предоставление услуг посредством моби...
1       3210  Плата за предоставление отчета по счету карты ...
2       3800  Плата за обслуживание банковской карты (за пер...
3       4000            Плата за получение наличных в Сбербанке
4       4001  Плата за получение наличных в Сбербанке (в дру...
..       ...                                                ...
150     2990                             Списание по требованию
151     2992  Списание средств для погашения задолженности п...
152     3001            Комиссия за обслуживание ссудного счета
153     3030  Плата за обслуживание банковской карты (за пос...
154     3120                Плата за досрочный перевыпуск карты

[155 rows x 2 columns]

In [173]:
transactions

customer_id   tr_datetime  mcc_code  tr_type    amount   term_id
0          39026145    0 10:23:26      4814     1030  -2245.92       NaN
1          39026145    1 10:19:29      6011     7010  56147.89       NaN
2          39026145    1 10:20:56      4829     2330 -56147.89       NaN
3          39026145    1 10:39:54      5499     1010  -1392.47       NaN
4          39026145    2 15:33:42      5499     1010   -920.83       NaN
...             ...           ...       ...      ...       ...       ...
999995     98801178  269 10:29:06      5921     1010 -23016.59    127873
999996     98801178  270 10:26:45      5211     1010 -61380.88  10114209
999997     98801178  270 15:18:53      5921     1010 -11800.27  10105341
999998     98801178  271 12:47:13      5211     1010 -16170.59  10114209
999999     98801178  271 14:51:56      5921     1010 -10025.77    123991

[1000000 rows x 6 columns]

In [174]:
gender_train

customer_id  gender
0        10928546       1
1        69348468       1
2        61009479       0
3        74045822       0
4        27979606       1
...           ...     ...
8395     90417572       0
8396     66837341       0
8397     10758984       1
8398     11376556       0
8399     29190649       0

[8400 rows x 2 columns]

## Описание данных
### Таблица ```transactions.csv```
##### Описание
Таблица содержит историю транзакций клиентов банка за один год и три месяца.

##### Формат данных

```
customer_id,tr_datetime,mcc_code,tr_type,amount,term_id
111111,15 01:40:52,1111,1000,-5224,111111
111112,15 15:18:32,3333,2000,-100,11122233
...
```
##### Описание полей

 - ```customer_id``` — идентификатор клиента;
 - ```tr_datetime``` — день и время совершения транзакции (дни нумеруются с начала данных);
 - ```mcc_code``` — mcc-код транзакции;
 - ```tr_type``` — тип транзакции;
 - ```amount``` — сумма транзакции в условных единицах со знаком; ```+``` — начисление средств клиенту (приходная транзакция), ```-``` — списание средств (расходная транзакция);
 - ```term_id``` — идентификатор терминала;


### Таблица ```gender_train.csv```

##### Описание
Данная таблица содержит информацию по полу для части клиентов, для которых он известен. Для остальных клиентов пол неизвестен.

##### Формат данных
```
customer_id,gender
111111,0
111112,1
...
```

##### Описание полей
 - ```customer_id``` — идентификатор клиента;
 - ```gender``` — пол клиента;

### Таблица ```tr_mcc_codes.csv```

##### Описание
Данная таблица содержит описание mcc-кодов транзакций.

##### Формат данных
```
mcc_code;mcc_description
1000;словесное описание mcc-кода 1000
2000;словесное описание mcc-кода 2000
...
```

##### Описание полей
 - ```mcc_code``` – mcc-код транзакции;
 - ```mcc_description``` — описание mcc-кода транзакции.

### Таблица ```tr_types.csv```

##### Описание
Данная таблица содержит описание типов транзакций.

##### Формат данных
```
tr_type;tr_description
1000;словесное описание типа транзакции 1000
2000;словесное описание типа транзакции 2000
...
```

##### Описание полей
 - ```tr_type``` – тип транзакции;
 - ```tr_description``` — описание типа транзакции;

#### Task 5.
1. В tr_types выберите произвольные 100 строк с помощью метода sample (указав при этом random_seed равный 242)
2. В полученной на предыдущем этапе подвыборке найдите долю наблюдений (стобец tr_description), в которой содержится подстрока 'плата' (в любом регистре). (\*)

Выведите ответ в виде вещественного числа, округлённого до двух знаков после запятой, отделив дробную часть точкой в формате "123.45"

##### Пояснения:
(\*) Строки "ПлатА за аренду", "ПлатАза аренду", "ПЛАТА" удовлетворяют условию, так как будучи переведёнными в нижний регистр содержат подстроку "плата".

In [238]:
np.random.seed(242)
tr_types_5 = tr_types.sample(100)
tr_types_5

tr_description
tr_type                                                   
6210                       Возврат покупки.POS Зарубеж. ТУ
7034     Перевод на карту/ с карты через АТМ (без взима...
2431     Перевод с карты на счет физ.лица в другом банк...
4001     Плата за получение наличных в Сбербанке (в дру...
8000                      Поправки по счету клиен. Зачисл.
...                                                    ...
4100                   Плата за получение наличных. Россия
2371     Списание с карты на карту по операции <перевод...
2421     Перевод с карты на счет физ.лица в другом тер....
4011     Плата за получение наличных в АТМ Сбербанка (в...
7040     Перевод на карту (с карты) через POS (в предел...

[100 rows x 1 columns]

In [320]:
tr_types_5.tr_description.str.contains('плата', case=False).sum() / len(tr_types_5)



0.26

#### Task 6. 
1. Для поля tr_type датафрейма transactions посчитайте частоту встречаемости всех типов транзакций tr_type в transactions.
2. Из перечисленных вариантов выберите те, которые попали в топ-5 транзакций по частоте встречаемости.

#### Выберите все верные пункты:
    - 1) Выдача наличных в АТМ Сбербанк России 
    - 2) Комиссия за обслуживание ссудного счета
    - 3) Списание по требованию	
    - 4) Оплата услуги. Банкоматы СБ РФ 
    - 5) Погашение кредита (в пределах одного филиала)
    - 6) Покупка. POS ТУ СБ РФ 

In [240]:
top5_tr_types = transactions.tr_type.value_counts().head(5)
top5_tr_types

1010    231117
2010    151166
7070    149006
1110    137658
1030    118975
Name: tr_type, dtype: int64

In [263]:
tr_types.loc[list(top5_tr_types.index)]

tr_description
tr_type                                                   
1010                                 Покупка. POS ТУ СБ РФ
2010                 Выдача наличных в АТМ Сбербанк России
7070     Перевод на карту (с карты) через Мобильный бан...
1110                                Покупка. POS ТУ Россия
1030                        Оплата услуги. Банкоматы СБ РФ

    - 1) Выдача наличных в АТМ Сбербанк России 
    - 4) Оплата услуги. Банкоматы СБ РФ 
    - 6) Покупка. POS ТУ СБ РФ 

#### Task 7.
1. В датафрейме transactions задайте столбец  customer_id в качестве индекса.
2. Выделите клиента с максимальной суммой транзакции (то есть с максимальным приходом на карту). (\*)
3. Найдите у него наиболее часто встречающийся модуль суммы приходов/расходов. (\**)


#### Выберите все верные пункты:
    - 1) 1122957.89
    - 2) 15721.41
    - 3) 22459.16
    - 4) 13475494.63
    - 5) 107407.78
    - 6) 65019.26

##### Пояснения:
(\*) Если у клиента были транзакции [-10000, 10, 0, -10], то максимумом будет являться значение 10. <br>
(\**) Если у клиента были транзакции [-10000, 10, 0, -10], то наиболее встречающийся модуль суммы транзакций равен 10, и встретился он 2 раза.

In [270]:
transactions_7 = transactions.set_index(['customer_id'])
transactions_7

tr_datetime  mcc_code  tr_type    amount   term_id
customer_id                                                     
39026145       0 10:23:26      4814     1030  -2245.92       NaN
39026145       1 10:19:29      6011     7010  56147.89       NaN
39026145       1 10:20:56      4829     2330 -56147.89       NaN
39026145       1 10:39:54      5499     1010  -1392.47       NaN
39026145       2 15:33:42      5499     1010   -920.83       NaN
...                   ...       ...      ...       ...       ...
98801178     269 10:29:06      5921     1010 -23016.59    127873
98801178     270 10:26:45      5211     1010 -61380.88  10114209
98801178     270 15:18:53      5921     1010 -11800.27  10105341
98801178     271 12:47:13      5211     1010 -16170.59  10114209
98801178     271 14:51:56      5921     1010 -10025.77    123991

[1000000 rows x 5 columns]

In [275]:
transactions_7.amount.idxmax()

26782922

In [287]:
amounts_of_customer = transactions_7.loc[transactions_7.amount.idxmax()].amount.abs()
amounts_of_customer

customer_id
26782922     31442.82
26782922     11935.25
26782922     11229.58
26782922     67377.47
26782922    129913.87
              ...    
26782922    988202.94
26782922     80852.97
26782922     22459.16
26782922     10488.43
26782922    116787.62
Name: amount, Length: 294, dtype: float64

In [288]:
amounts_of_customer.mode()

0    22459.16
Name: amount, dtype: float64

#### Task 8.
1. Найдите максимальную разницу между парой из следующих значений (\*):
    - Медиана суммы транзакций
    - Медиана суммы транзакций по тем строкам, которые ни в одном из своих столбцов не содержат пустые значения
    - Медиана суммы упорядоченного по столбцу amount датафрейма transactions, у которого удалены дублирующиеся по столбцам [mcc_code, tr_type] строки, причём при удалении соответстующих дублей остаются только последние из дублирующихся строк (keep='last')

Выведите ответ в виде вещественного числа, округлённого до двух знаков после запятой, отделив дробную часть точкой в формате "123.45"

##### Пояснения:
(\*) Для вычисления максимальной разницы между значениями списка можно использовать функцию np.ptp  
(\**) Если в результате получились значения [1,3,5], то макимальная разница между ними 4 == 5-1.

In [307]:
transactions

customer_id   tr_datetime  mcc_code  tr_type    amount   term_id
0          39026145    0 10:23:26      4814     1030  -2245.92       NaN
1          39026145    1 10:19:29      6011     7010  56147.89       NaN
2          39026145    1 10:20:56      4829     2330 -56147.89       NaN
3          39026145    1 10:39:54      5499     1010  -1392.47       NaN
4          39026145    2 15:33:42      5499     1010   -920.83       NaN
...             ...           ...       ...      ...       ...       ...
999995     98801178  269 10:29:06      5921     1010 -23016.59    127873
999996     98801178  270 10:26:45      5211     1010 -61380.88  10114209
999997     98801178  270 15:18:53      5921     1010 -11800.27  10105341
999998     98801178  271 12:47:13      5211     1010 -16170.59  10114209
999999     98801178  271 14:51:56      5921     1010 -10025.77    123991

[1000000 rows x 6 columns]

In [314]:
m1 = transactions.amount.median()
m1

-5614.79

In [316]:
transactions_8_2 = transactions.dropna()
m2 = transactions_8_2.amount.median()
m2

-5596.6

In [317]:
transactions_8_3 = transactions.sort_values(by='amount')
transactions_8_3.drop_duplicates(['mcc_code', 'tr_type'], keep='last', inplace=True)
m3 = transactions_8_3.amount.median()
m3

-920.83

In [319]:
np.ptp([m1, m2, m3])

4693.96

1) 1.18

2) Серия (1) совпадает с серией (2), так как в каждом из случаев серия создаётся из списка строк

3) Ответ (1) - не может быть определён / Ответ (2) - 57591.19

4) -15.15,51

5) 0.26

6)
    - 1) Выдача наличных в АТМ Сбербанк России 
    - 4) Оплата услуги. Банкоматы СБ РФ 
    - 6) Покупка. POS ТУ СБ РФ 

7) 22459.16

8) 4693.96